src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+static int
+get_rebuild_target(struct pool_map *pmap, struct pool_target **target,
+		   uint64_t key, uint8_t *dom_used,
+		   struct pl_obj_layout *layout, struct daos_obj_md *md)
+{
+	uint8_t                 *used_tgts = NULL;
+	uint32_t                selected_dom;
+	uint32_t                fail_num = 0xFFc5;
+	uint32_t                try = 0;
+	struct pool_domain      *target_selection;
+	struct pool_domain      *root;
+	int rc = 0;
+
+	rc = pool_map_find_domain(pmap, PO_COMP_TP_ROOT, PO_COMP_ID_ALL, &root);
+	if (rc == 0) {
+		D_ERROR("Could not find root node in pool map.");
+		rc = -DER_NONEXIST;
+		return rc;
+	}
+
+	while (1) {
+
+		uint8_t skiped_targets = 0;
+		uint16_t num_bytes = 0;
+		uint32_t num_doms;
+
+		num_doms = root->do_child_nr;
+
+		uint64_t child_pos = (uint64_t)(root->do_children)
+				     - (uint64_t)root;
+		child_pos = child_pos / sizeof(struct pool_domain);
+
+		/*
+		 * If all of the nodes have been used for shards but we
+		 * still have shards to place mark all nodes as unused
+		 * so duplicates can be chosen
+		 */
+		if (isset_range(dom_used, child_pos, child_pos +
+				 num_doms - 1)) {
+			clrbit_range(dom_used, child_pos,
+				(child_pos + (num_doms - 1)));
+		}
+
+		/*
+		 * Choose domains using jump consistent hash until we find a
+		 * suitable domains that has not already been used.
+		 */
+		do {
+			key = crc(key, fail_num++);
+			selected_dom = jump_consistent_hash(key, num_doms);
+			target_selection = &(root->do_children[selected_dom]);
+		} while (isset(dom_used, (selected_dom + child_pos)));
+
+		/* Mark this domain as used */
+		setbit(dom_used, (selected_dom + child_pos));
+
+		/* To find rebuild target we examine all targets */
+		num_doms = target_selection->do_target_nr;
+
+		num_bytes = (num_doms / 8) + 1;
+
+		D_ALLOC_ARRAY(used_tgts, num_bytes);
+		if (used_tgts == NULL)
+			return -DER_NOMEM;
+
+		/*
+		 * Add the initial layouts targets to the bitmap of checked
+		 * targets.
+		 */
+		int i;
+		uint64_t start_pos;
+		uint64_t end_pos;
+
+		start_pos = (uint64_t)(&((target_selection->do_targets[0])));
+		end_pos = start_pos + (num_doms * sizeof(**target));
+
+		for (i = 0; i < layout->ol_nr; ++i) {
+			int id = layout->ol_shards[i].po_target;
+			uint64_t position;
+
+			pool_map_find_target(pmap, id, target);
+			position = (uint64_t)(*target);
+
+
+			if (position >= start_pos && position < end_pos) {
+				setbit(used_tgts, ((position - start_pos)
+				/ sizeof(**target)));
+				skiped_targets++;
+			}
+		}
+
+
+		/*
+		 * Attempt to choose a fallback target from all targets found
+		 * in this top level domain.
+		 */
+		do {
+			key = crc(key, try++);
+
+			selected_dom = jump_consistent_hash(key, num_doms);
+			*target = &target_selection->do_targets[selected_dom];
+
+			/*
+			 * keep track of what targets have been tried
+			 * in case all targets in domain have failed
+			 */
+			if (isclr(used_tgts, selected_dom))
+				skiped_targets++;
+
+			if (pool_target_unavail(*target))
```

Comment:
If the alternative target is down as well, but if it has higher fseq than the current target to be remapped, we need to stop remapping for the current one, the current one should be removed from the remap list, and new failed target with higher fseq will be inserted into remap list and being processed later. (The remap list processing should strictly follow the fseq order)

Reply:
correct. This code has been removed and now should be returned based on

> !pool_target_unavail(*target) || (*target)->ta_comp.co_fseq > md->omd_ver